### PR TITLE
Allow different modes for deepMerge

### PR DESCRIPTION
**Pull request by halfvim**
When performing a deepMerge, allow user to specify a merge mode, which can be
- Default: treats Array as value, similar to Null, Boolean, String or Number
- Index: replace the elements in this JSON array with the elements in the argument JSON at corresponding position.
- Concat: concatenate the elements in this JSON array and the argument JSON array.

This PR handles issue #271

### CODE REVIEW

File: modules/core/shared/src/main/scala/io/circe/Json.scala [hunk 1]
```diff
@@ -228,6 +257,14 @@ sealed abstract class Json extends Product with Serializable {
 
 object Json {
 
+  /**
+   * Merge strategy used for deepMerge operation when merging two arrays.
+   */
+  object MergeMode extends Enumeration {
```
Comment (travisbrown) [inline modules/core/shared/src/main/scala/io/circe/Json.scala]: I think it'd be better to use the `sealed trait` + `case object` pattern here, for consistency.

Comment (halfvim) [reply modules/core/shared/src/main/scala/io/circe/Json.scala]: > I think it'd be better to use the `sealed trait` + `case object` pattern here, for consistency.

done.

**Comment by halfvim** [comment]
not sure what exactly to be done for moving stuff to `io.circe.extras`, but i've changed enum to case objects.

**Comment by JD557** [comment]
Sorry for commenting out of nowhere, but wouldn't it be better to have something like `def deepMerge(that: Json, mergeLeafs: (Json, Json) => Json = _._2)`?

The current implementation only cares about arrays, but there are other cases where setting a merging strategy can be useful (e.g. when merging two numbers, sum the result).

**Comment by ShahOdin** [comment]
@nrthyrk Just curious why this was closed?

**Comment by jcallin** [comment]
Yeah would love to have this merged

**Pull request closed without merging**